### DISCUSSION TITLE
Custom objective questions

### DISCUSSION THREAD

**Discussion by unary-code** [MilesCranmer/PySR#375] [Q&A]
I can not find a reason for the difference between the output of objective function using Julia helper and output of objective function using model.

The math function isPySRFunction(X=>11.9731872446037).
the model.equations_.iloc[3, :] is
```
complexity                                                      15
loss                                                       0.08352
score                                                     0.217383
equation         '(((((((0.3079866 / 0.026352234) + 0.9724522) - 0.3079866) + -0.18929175) + x0) + -0.18929175) - x0)
sympy_format                                      11.9731872446037
lambda_format                    PySRFunction(X=>11.9731872446037)
Name: 3, dtype: object

```

The model's reported loss is 0.08352. This is incorrect, given that the math function is 11.9731872 (I don't want the model to return this value).

While when I run the objective function using julia helper object, I get a loss of 1.0000000015930166. This is correct.

Concrete yes/no question you can answer:
Does the model call my custom objective function eval_loss multiple times, or just one time, in order to find its reported value of loss?

I can provide code, but the julia helper object ran the same code as the model's objective function.
Specifically, the julia helper object ran this code (and returned a final loss of 1.000000000001593):
```
jl.eval("""
# YOU USE THIS CODE BLOCK TO TEST OUT THE VALUE OF THE CUSTOM OBJECTIVE FUNCTION

using Distributions
using SymbolicRegression


function eval_loss_real(tree, dataset, options)
    MAX_THRESHOLD_ALLOWED = 10000
    

    prediction, flag = eval_tree_array(tree, dataset, options)
    !flag && return L(Inf)
    
    
    
    num_points = length(prediction)
    println("num_points is ", num_points)
    num_points = size(prediction)[1]
    println("num_points is ", num_points)
    
    
    
    
    # BELOW SECTION IS TO CHECK FOR IF THE GUESSED FUNCTION TAKES ON ANY POSITIVE INFINITY VALUES
    # OR NEGATIVE INFINITY VALUES ON THE DOMAIN X=[0, 1].
    
    
    # NOTE: reshape([x], 1, 1) should give same dimensions
    # as reshape([x;], 1, 1).
    # f(x) = x -> 1/(eval_tree_array(tree, reshape([Float32(x)], 1, 1), options))
    f_recip(x) = 1 / ((eval_tree_array(tree, reshape([Float32(x)], 1, 1), options))[1][1])
    
    # fp(x) = ForwardDiff.derivative(f,float(x))
    fp_recip(x) = ( -1 / ((eval_tree_array(tree, reshape([Float32(x)], 1, 1), options))[1][1])^2 * (eval_diff_tree_array(tree, reshape([Float32(x)], 1, 1), options, 1))[2][1])
    
    # 0.26047713
    #println(((eval_tree_array(tree, reshape([Float32(0.01)], 1, 1), options))[1][1])^2)
    #println(fp_recip(0.01))
    
    global found_a_recip_root = true
    global recip_roots = -3
    global recip_roots_2 = -4
    try
        global recip_roots = find_zero((f_recip, fp_recip), 0.3, Roots.Newton(), maxevals=100)
        global recip_roots_2 = find_zero((f_recip, fp_recip), 0.01, Roots.Newton(), maxevals=100)
        println("recip_roots is", recip_roots)
        println("recip_roots_2 is", recip_roots_2)
    catch
        global found_a_recip_root = false
        # "Convergence Failed"
    end
    
    #println("recip_roots is", recip_roots)
    
    
    # If the function f(x) represented by the tree "tree"
    # has a value f(x) close to positive infinity or negative infinity
    # when x is in [0, 1], then return L(Inf).
    if found_a_recip_root
        if (recip_roots >= 0 && recip_roots <= 1)
            return (Inf)
        end
    end
    
    if found_a_recip_root
        if (recip_roots_2 >= 0 && recip_roots_2 <= 1)
            return (Inf)
        end
    end
    
    
    # BELOW SECTION IS TO CHECK FOR IF THE GUESSED FUNCTION TAKES ON ANY NEGATIVE VALUES OR ZERO VALUES
    # ON THE DOMAIN X=[0, 1].
    f(x) = ((eval_tree_array(tree, reshape([Float32(x)], 1, 1), options))[1][1])
    
    fp(x) = ((eval_diff_tree_array(tree, reshape([Float32(x)], 1, 1), options, 1))[2][1])
    
    global found_a_root = true
    global roots = -3
    try
        global roots = find_zero((f, fp), 0.3, Roots.Newton(), maxevals=100)
        #println("roots is", roots)
    catch
        global found_a_root = false
        # "Convergence Failed"
    end
    
    println("roots is", roots)
    
    if found_a_root
        if (roots >= 0 && roots <= 1)
            println("found a root so returned L(Inf) and string rep of the tree was")
            println(string_tree(tree, options))
            return (Inf)
        end
    end
    
    
    
    global integral = 1
    err = -3
    try
        my_tuple = quadgk(x -> (eval_tree_array(tree, reshape([Float32(x)], 1, 1), options))[1][1], 0, 1, rtol=0.01)
        global integral = my_tuple[1]
        global err = my_tuple[2]
    catch
        println("A error when calling quadgk")
        return (Inf)
    end
    
    
    
    norm_constant = integral
    println("norm_constant is", norm_constant)
    if (integral <= 0)
        println("integral was <=0")
        return (Inf)
    end
    
    if (isinf(integral))
        println("integral was Inf")
        return (Inf)
    end
    
    actual_probs = (prediction) / norm_constant
    
    # length(actual_probs) equals length(prediction)
    
    # We really care about the EXPECTED VALUE OF L, where L = -1 * sum(log.(actual_probs)) / (length(prediction))
    # given that dataset.X is generated by the true probability distribution.
    # E(L) is really equal to the definite integral from x=0 to x=1 of ln(guessed_f(x)) * f_true(x) dx.
    # As num_points goes to +infinity, the variance of L becomes 0.
    
    
    final_loss = exp(-1 * sum(log.(actual_probs)) / (num_points))
    #println("final_loss is ", final_loss)
    return final_loss
end

options = SymbolicRegression.Options(
    binary_operators=[+, *, /, -, ^],
    unary_operators=[cos, sin],
    enable_autodiff=true
)

num_points = 100

x = rand(Beta(3, 2), num_points)
x = reshape(x, 1, num_points)
x = Float32.(x)

x1 = Node(;feature=1)

tree = Node(;val=Float32(11.9731872446037))
tree = Node{Float32}(tree)

println(string_tree(tree, options))

my_loss = eval_loss_real(tree, x, options)
println("FINAL LOSS was ", my_loss)


""")
```

**Comment by MilesCranmer** [comment]
> Does the model call my custom objective function eval_loss multiple times, or just one time, in order to find its reported value of loss?

Only one time per expression.

**Reply by unary-code** [reply]
Okay, this is confusing to me. Since I used the same exact code in the objective function for the model and in the function that I called directly using jl.eval(""" """), and the model's reported loss is inaccurate and different than the loss I get when I call the objective function directly using jl.eval(""" """).

Perhaps I can send you a ipynb file or a txt file with the barebones code (no code that is irrelevant) and you can run it yourself? What file extension type would you prefer?

**Reply by MilesCranmer** [reply]
Maybe try running this directly from Julia (`Pkg.add("SymbolicRegression"); using SymbolicRegression`)? Also, you might want to pass it a `Dataset` object rather than an `Array` as is being done now, so as to ensure absolute consistency in both cases?

**Reply by unary-code** [reply]
For example, the model's reported loss for the function "((((x0 + x0) + x0) + x0) - x0)" which is really 3*x0 was: "loss                                   0.600895", when in reality I would expect a loss of 0.8960009551.

This is because my model calculates the norm constant (integral of ((((x0 + x0) + x0) + x0) - x0) from x=0 to x=1, which is 3/2), which should be a norm constant of 3/2. Then divides the guessed function f_hat by the norm constant.
Then takes the mean of ln of (f_hat(x_i) / norm_constant) for every point x_i in the dataset, then takes the negative of that, then returns a loss of e to that power.
The negative value of the definite integral of ln(2x) * 12 * x^2 * (1-x) from x=0 to x=1  is −0.109813847227 and e to that power is 0.896000955143.
Where  12 * x^2 * (1-x) is the  Beta(alpha=3, beta=2) distribution, so I expect the objective function's returned loss to be around 0.896 .

Perhaps my randomly-generated dataset X (generated from the Beta(alpha=3, beta=2) distribution) that I pass into model.fit(X, y) is too small, and thus there is some variance. X is a numpy matrix and X has 100 datapoints, and X's shape is (100, 1).
y's shape is (100,) but that's irrelevant because my objective function does not use dataset.y

**Comment by unary-code** [comment]
> Maybe try running this directly from Julia (`Pkg.add("SymbolicRegression"); using SymbolicRegression`)? Also, you might want to pass it a `Dataset` object rather than an `Array` as is being done now, so as to ensure absolute consistency in both cases?

I already did run
```
jl.eval("""
Pkg.add("SymbolicRegression")
using SymbolicRegression
""")
```
before I ran my objective function using the julia helper "jl".

That's a good point: Now I'll try changing the objective function's inputs to be a dataset so it is consistent with the model's objective function. That was the only difference between my Julia helper's objective function and the model's objective function.

**Reply by MilesCranmer** [reply]
I know but I'm just saying maybe run this all in Julia directly, rather than through PyJulia? Once everything is working you could move back to PySR but if one is doing that much customization it perhaps might be easier to work from Julia itself.
